Fix regressions in some video modes' aspect and line-counts

kcgen opened the pull request:
This is more cleanup from the vgamode integration PR. Changes are committed per-mode to aid in bisecting if needed.

Please test extensively (with anything you want to throw at it!) to help catch any *new* regressions. 

Things to try:

- (S)VGA, EGA, CGA, and Tandy machine types.
- When using VGA machine types, confirm sub-350 line modes are double-scanned or single-scanned if `[dosbox] force_vga_single_scan` is `true`.
- Confirm aspect is correct (or not double-wide or too tall).
- When using EGA/CGA/Tandy machine types, confirm lines are "thick n' chunky", and not double-scanned. 
- Confirm 350+-line modes are never double scanned (in any mode).
- Confirm refresh rates are logged correctly.

johnnovak commented:
I will test and confirm with all screen modes known to mankind 😎 ...in the evening, I'm at work today.

johnnovak commented:
Outstanding work @kcgen! Everything works absolutely correctly now; specifically, all CGA and EGA modes are now double-scanned correctly on VGA, `force_vga_single_scan` is 100% correct, and all emulated pre-VGA adapters use single-scanning 100% of the time! Excellent! 🎉 

I'm so happy you managed to get this perfect! The lack of correct handling of double-scanning has been a long-standing sore point of mine! ❤️ 🚀 🎊 

Going through the checklist in detail (Deluxe Paint II is very handy by the way to check every resolution quickly):

✅ (S)VGA, EGA, CGA, and Tandy machine types.
✅ When using VGA machine types, confirm sub-350 line modes are double-scanned or single-scanned if [dosbox] force_vga_single_scan is true.
✅ Confirm aspect is correct (or not double-wide or too tall)
✅ When using EGA/CGA/Tandy machine types, confirm lines are "thick n' chunky", and not double-scanned.
✅ Confirm 350+-line modes are never double scanned (in any mode).
⚔️ Confirm refresh rates are logged correctly.

There are a few small issues with the logging only in a minority of cases:

## VGA 

I think VGA not only line-doubled 320x200, but also double the vertical resolution. Reference: http://nerdlypleasures.blogspot.com/2013/10/320x200-resolution-of-choice-for-ibm-pc.html

> The high resolution display supported a 31.5KHz scan rate and 70Hz vertical refresh rates for all VGA modes, including emulated modes, except for the 640x480 mode, which used 60Hz.  200-line modes would be double scanned, with each pixel being double-clocked and each vertical line being repeated to fill up the refresh rate.  This gives a different kind of scan-line structure compared with earlier monitor.

So at the very least the log should say 640x400 for mode 13h. Technically, we should double vertically too, although probably it makes little practical difference...

```
DISPLAY: VGA 320x400 8-bit (mode 13h) at 70.086 Hz VFR, scaled to 2205x1654 with 0.6001 pixel aspect ratio
```

Same for the CGA and EGA 200-line modes on VGA; vertical res should be 640:

```
DISPLAY: CGA 320x400 4 color (mode 04h) at 70.086 Hz CFR, scaled to 2880x2160 with 0.6 pixel aspect ratio
```
```
DISPLAY: EGA 320x400 16 color (mode 0Dh) at 70.086 Hz CFR, scaled to 2880x2160 with 0.6 pixel aspect ratio
```


## CGA

Issue: just the logging is incorrect; should say "CGA" instead of "Tandy
```
DISPLAY: Tandy 320x200 4 color (mode 04h) at 59.923 Hz VFR, scaled to 2205x1654 with 1.2 pixel aspect ratio
```
Issue: just the logging is incorrect; should say "CGA" instead of "Tandy
```
DISPLAY: Tandy 640x200 2 color (mode 06h) at 59.923 Hz VFR, scaled to 2205x1654 with 2.4 pixel aspect ratio
```

## MCGA

Issue: just the logging is incorrect; should say "MCGA" instead of "VGA"
```
DISPLAY: VGA 320x400 8-bit (mode 13h) at 70.086 Hz VFR, scaled to 2205x1654 with 0.6001 pixel aspect ratio
```
Issue: just the logging is incorrect; should say "MCGA" instead of "VGA"
```
DISPLAY: VGA 640x480 monochrome (mode 11h) at 59.94 Hz VFR, scaled to 2205x1654 with 1 pixel aspect ratio
```

johnnovak reviewed (approved):
Nice to see things actually got simpler in the code! 🎉

GranMinigun commented:
Dangerous Dave in the Haunted Mansion shows interesting line artefacts right at the beginning (just jump on the tree). Bisection points at 9f2f833f59c6418b2bf69f6ec0ad14bb7b7ef10b, and this PR, unfortunately, doesn't fix the issue. Is that in scope or something for a separate PR?

ThomasEricB commented:
> Dangerous Dave in the Haunted Mansion shows interesting line artefacts right at the beginning (just jump on the tree). Bisection points at [9f2f833](https://github.com/dosbox-staging/dosbox-staging/commit/9f2f833f59c6418b2bf69f6ec0ad14bb7b7ef10b), and this PR, unfortunately, doesn't fix the issue. Is that in scope or something for a separate PR?

@kcgen Did ended up borrowing @GranMinigun 's Lambda code from https://github.com/dosbox-staging/dosbox-staging/commit/1453a5b657ecee5879880dc09c55a4da87230a53 ? I can't find the commit... Perhaps that has something to do with this?

GranMinigun commented:
No, it's unrelated.

kcgen commented:
@johnnovak , @GranMinigun -- thanks for all the testing!

> at the very least the log should say 640x400 for mode 13h.

The vertical line count is being double-scanned up from 200 to 400, but we still have 320 columns (or 320 pixels in width):

```
DISPLAY: VGA 320x400 8-bit (mode 13h)
```

> Technically, we should double vertically too, although probably it makes little practical difference...

It would make a difference. Going frmo from 320-pixels wide up to 640, shaders will now see two pixels horizontally instead of one. It would be similar to double-scanning vertically as well (which I don't think we want).

> Same for the CGA and EGA 200-line modes on VGA; vertical res should be 640

The same issue would happen here too; width would double from 320 to 640, and shaders would then operate on each pixel (now twice as many) across the width.

> CGA logging

Fixed - thanks!

> "MCGA" mono instead of "VGA"

:+1: , fixed 

> MCGA 320x400 8-bit (mode 13h)

This mode is identical to VGA 320x200 13h.. I have no way of differentiating (we don't have `machine = mcga`)

> Dangerous Dave in the Haunted Mansion shows interesting line artefacts right at the beginning (just jump on the tree).

Great catch @GranMinigun and thanks for bisecting - should be fixed (and now further simplified that function).

johnnovak commented on src/hardware/vga.cpp:
```diff
@@ -65,12 +69,12 @@ std::pair<const char*, const char*> VGA_DescribeMode(const VGAModes video_mode_t
-	case M_TANDY16:            return {"Tandy", "16 color"};
+	case M_TANDY2:             return {maybe_cga_on_tandy(), "2 color"};
+	case M_TANDY4:             return {maybe_cga_on_tandy(), "4 color"};
+	case M_TANDY16:            return {maybe_cga_on_tandy(), "16 color"};
```
Nice! 😎

johnnovak commented on src/hardware/vga.cpp:
```diff
@@ -65,12 +69,12 @@ std::pair<const char*, const char*> VGA_DescribeMode(const VGAModes video_mode_t
 	case M_EGA: // see comment below
 	    switch (video_mode_id) {
-	    case 0x011:            return {"VGA",   "monochrome"};
+	    case 0x011:            return {"MCGA",  "monochrome"};
```
👍🏻

johnnovak commented:
> It would make a difference. Going from 320-pixels wide up to 640, shaders will now see two pixels horizontally instead of one. It would be similar to double-scanning vertically as well (which I don't think we want).

But that's how the hardware works.

kcgen commented:
> But that's how the hardware works. If we want to do the same thing, we need to double vertically too.
My understanding is VGA outputs a 640x400 image for 320x200. 

👍 Will do!

GranMinigun commented:
Confirming Dave now works through mansion just fine.

kcgen commented:
> My understanding is VGA outputs a 640x400 image for 320x200. So effectively it 2x integer upscales in hardware.

> Here we go, although it's a bit confusing, but my reading of it is that the pixel clock is set to 640 pixels per line, and the colour output is "latched" at every second pixel
https://www.vogons.org/viewtopic.php?p=816033#p816033

Was able to add width-doubling at the draw level for VGA and EGA modes.

I wasn't able to get CGA modes width-doubling, so instead I use aspect correction when rendering to achieve it; however, the sub-350-line doubling is still working. Assuming your tests show there is no visible difference:

> I'll run a few tests later to see what observable difference 640x400 vs 320x400 makes when using CRT shaders (the difference is zero with the sharp shader).

... I've included a simple log work-around (with TODO and description) so at least we log the mode "as seen" by the user to avoid confusion (and log what /should/ be happening at the VGA draw level instead of at the rendering level).

kcgen commented:
I can't see anything left. @johnnovak and @GranMinigun : if you can give it a final pass, I think we're ready for merge.

ThomasEricB commented:
@kcgen We are not ready to merge. Found a very grave bug on Tandy.

ThomasEricB reviewed (changes requested):
@kcgen We are not ready to merge. Found a very grave bug on Tandy. In a couple of minutes I'll be explaining it in more detail

ThomasEricB reviewed (approved):
Oops. I left @johnnovak 's fix video shader on by default. Forget what I just said

johnnovak commented:
> > My understanding is VGA outputs a 640x400 image for 320x200. So effectively it 2x integer upscales in hardware.
> 
> > Here we go, although it's a bit confusing, but my reading of it is that the pixel clock is set to 640 pixels per line, and the colour output is "latched" at every second pixel
> > https://www.vogons.org/viewtopic.php?p=816033#p816033
> 
> Was able to add width-doubling at the draw level for VGA and EGA modes.
> 
> I wasn't able to get CGA modes width-doubling, so instead use aspect to achieve it at the rendering level. (sub-350 line doubling is still working and unaffected) Assuming your tests show there is no visible difference ...
> 
> > I'll run a few tests later to see what observable difference 640x400 vs 320x400 makes when using CRT shaders (the difference is zero with the sharp shader).
> 
> ... I've included a simple log work-around (with TODO and description) so we log the mode "as seen" by the user to avoid confusion (and log what /should/ be happening at the VGA draw level instead of at the rendering level).

Great. I'll test all that later today.

johnnovak commented:
Here we go @kcgen:

## General comments

- I don't think this is working (more details below): "Was able to add width-doubling at the draw level for VGA and EGA modes." 
- 256-colour modes get logged as "8-bit" colour. This is because technically they get rendered as 24-bit RGB, which gets logged incorrectly as "8-bit". This needs to be changed to "24-bit". That's fine for the "true 24-bit VESA modes", but for the 256-colour modes we should log the _original_ colour depth, not just simply whatever pixel format the actual surface has—that's an implementation detail.
- I suggest to log pixel aspect ratios as "1:1" and "1:1.2" instead of "1" and "1.2", etc. [Pixel aspect ratios](https://en.wikipedia.org/wiki/Pixel_aspect_ratio) are fractions, we just happen to log the "normalised" version where the numerator (first number) is always 1.
- I'd like to have "color" changed to "colour" in the logs 😄 
- I've tested every possible standard resolution & video card combo, plus forcing single scanlines on VGA. What I haven't listed below is fine, so basically only the line-doubled 320px wide modes are still not correct (they're not pixel-doubled).

## machine = vga

### 320x200 / 4-colour

No pixel-doubling, only line-doubling.

```
DISPLAY: CGA 320x400 4 color (mode 04h) at 70.086 Hz VFR, scaled to 2205x1654 with 0.6001 pixel aspect ratio
```

### 320x200 / 16-colour

No pixel-doubling, only line-doubling.

```
DISPLAY: EGA 320x400 16 color (mode 0Dh) at 70.086 Hz VFR, scaled to 2205x1654 with 0.6001 pixel aspect ratio
```

### 320x200 / 256-colour

No pixel-doubling, only line-doubling.

```
DISPLAY: VGA 320x400 8-bit (mode 13h) at 70.086 Hz VFR, scaled to 2205x1654 with 0.6001 pixel aspect ratio
```

ThomasEricB commented:
> Everything works very nicely now @kcgen without any issues, well done! 🎉 I've tested all standard resolutions on all emulated cards (incl. forced single scanlines on VGA), plus all the Quake custom resolutions. 

Can confirm what @johnnovak has said. So far, no issues found on my end. I think this should be safe to merge into main.

> I'd like to have "color" changed to "colour" in the logs 😄

This is my favo**u**rite take that @johnnovak ever made 😄

johnnovak commented:
> This is my favo**u**rite take that @johnnovak ever made 😄

Hey, I thought the deinterlace shader was your favourite! 😛

ThomasEricB commented:
> Hey, I thought the deinterlace shader was your favourite! 😛

Nop, that's my **favorite**!

bitstrom commented:
Pinball Fantasies and Pinball Illusions are not working properly if `[dosbox]` `force_vga_single_scan` is `true`.
Parts of the screen are missing.

kcgen commented:
@johnnovak : updated the  colo*u*r bit-depth and pixel-aspect logging:

```
Text 720x400 16-colour (mode 03h) at 70.087 Hz synced CFR,
scaled to 2132x1599 with 1:1.35 pixel aspect ratio
```

```
VGA 320x200 256-colour from 6-bit RGB palette (mode 13h) at 70.086 Hz VFR, scaled to 2132x1599 with 1:1.2 pixel aspect ratio
```

kcgen commented:
This wouldn't be complete without some screenshots!

Fortunately no regressions from upstream.

![cdshock_001](https://user-images.githubusercontent.com/1557255/236692619-9f9a0920-efe8-4693-95c1-cfeb42581973.png)

![cdshock_000](https://user-images.githubusercontent.com/1557255/236692620-d79ac25c-bd98-4dd8-98e8-a59c0574e2fe.png)

![cdshock_003](https://user-images.githubusercontent.com/1557255/236692621-8abb29b8-3eee-4d37-adfe-6b86b46633f2.png)

![cdshock_002](https://user-images.githubusercontent.com/1557255/236692622-1cbae6cc-edbd-4820-b32b-87b953958555.png)

![entrap_000](https://user-images.githubusercontent.com/1557255/236692624-a4ddb95e-d835-493a-856d-5132c3ac45d2.png)

![entrap_001](https://user-images.githubusercontent.com/1557255/236692626-1ceeae97-4135-46a6-8b97-7d53cf53a906.png)

![entrap_002](https://user-images.githubusercontent.com/1557255/236692627-2935a0f7-75b4-40a0-afb6-4cfc4c098e37.png)

![entrap_003](https://user-images.githubusercontent.com/1557255/236692628-faadc66e-1ea1-45c5-ad13-10c686000792.png)

![entrap_004](https://user-images.githubusercontent.com/1557255/236692629-1b333484-aa14-4d3c-9c3d-8ca27e92e960.png)

![illusion_000](https://user-images.githubusercontent.com/1557255/236692630-184dcb4a-40ed-4bc1-a9f9-9f25e98862e9.png)

![intro_000](https://user-images.githubusercontent.com/1557255/236692632-d3f6e9c2-3c47-4ae4-9e9d-e50e4861bc5c.png)

![intro_001](https://user-images.githubusercontent.com/1557255/236692633-8fa96538-36d0-4089-9502-2524438cf511.png)

![scorch_001](https://user-images.githubusercontent.com/1557255/236692634-c54d2f89-e87f-4cf8-aa5d-c0dc66e72c18.png)

![scorch_000](https://user-images.githubusercontent.com/1557255/236692635-d466fd3f-2c65-42bf-85cd-97092e1ac034.png)

![scorch_006](https://user-images.githubusercontent.com/1557255/236692637-ec729858-5c85-44f6-952c-fea3685f3ed0.png)

![scorch_005](https://user-images.githubusercontent.com/1557255/236692638-7ca504ea-1f30-493b-8697-7f7011076f42.png)

![scorch_004](https://user-images.githubusercontent.com/1557255/236692640-e6602c0a-3dbe-4531-98f8-e327555e9c25.png)

![scorch_002](https://user-images.githubusercontent.com/1557255/236692641-4c6b55f2-5aca-4770-9b47-7e1eddd97c8a.png)

![scorch_003](https://user-images.githubusercontent.com/1557255/236692643-7665390b-b11f-4e05-8212-b090209fe774.png)

![table4_000](https://user-images.githubusercontent.com/1557255/236692645-53a97057-5d2b-4334-b471-b9c08e3bbeb6.png)

![2023-05-07_09-23](https://user-images.githubusercontent.com/1557255/236692669-07dfcc35-f0ae-46a0-9426-e42b43822378.png)

![2023-05-07_09-22](https://user-images.githubusercontent.com/1557255/236692671-a873e517-888b-43c8-81b4-6c830a4a3598.png)

![2023-05-07_09-21](https://user-images.githubusercontent.com/1557255/236692672-4425bce5-f8b3-414d-b1fc-8a76d8507fd9.png)

![2023-05-07_09-20](https://user-images.githubusercontent.com/1557255/236692673-1ce632f5-3ea0-45d1-a7ff-95284a206acb.png)

![2023-05-07_09-17](https://user-images.githubusercontent.com/1557255/236692674-62e3c636-db9c-4676-aa36-a6c8724e5b73.png)

![2023-05-07_09-15](https://user-images.githubusercontent.com/1557255/236692675-94672728-ae55-4cf1-888b-5f0688ed5efd.png)

![2023-05-07_09-11](https://user-images.githubusercontent.com/1557255/236692676-7b6b4430-551b-4982-bc3e-1a7e84da68e8.png)

![2023-05-07_09-10](https://user-images.githubusercontent.com/1557255/236692677-2c4e7893-3243-47d9-b521-dd24330b47cb.png)

![2023-05-07_09-07](https://user-images.githubusercontent.com/1557255/236692678-8ab85d0d-15b0-4223-92df-a2b1b9a5e1f8.png)

![2023-05-07_09-05](https://user-images.githubusercontent.com/1557255/236692679-25568af9-809d-4beb-9272-cedb0c64a48f.png)

![2023-05-07_09-04](https://user-images.githubusercontent.com/1557255/236692681-9321889b-09f7-41ea-923c-dad8377376c1.png)

![2023-05-07_09-02](https://user-images.githubusercontent.com/1557255/236692682-4621d91d-7b22-4489-9180-33149cf60156.png)

![2023-05-07_08-59](https://user-images.githubusercontent.com/1557255/236692683-5e421ae1-b6c9-41b6-98eb-3683b395dcc9.png)

![2023-05-07_08-55](https://user-images.githubusercontent.com/1557255/236692684-a153e9ae-d9ed-4a64-89a7-617776ed9d44.png)

![2023-05-07_08-42](https://user-images.githubusercontent.com/1557255/236692685-aed45dd2-f647-46fd-9a30-0b90ab2a9306.png)

![2023-05-06_19-30](https://user-images.githubusercontent.com/1557255/236692686-761effac-4291-421b-94b8-9a41998fc23d.png)

![2023-05-06_19-26](https://user-images.githubusercontent.com/1557255/236692690-d282aa6f-2c9d-49e3-b9af-818ffb607e14.png)

![2023-05-06_19-24](https://user-images.githubusercontent.com/1557255/236692691-5ecd7ed6-9b7d-42ed-985f-279e6b6b5bf8.png)

kcgen commented:
Ready for (hopefully) the last round of reviews :-)

bitstrom commented:
There is still some weirdness in Pinball Fantasies and Pinball Illusions.

Pinball Fantasies:
`force_vga_single_scan` is `false`

When i let the intro run without interruption the pinball select screen colours are wrong and this is logged.
`DISPLAY: VGA 640x480 16-colour from 6-bit RGB palette (mode 12h) at 59.713 Hz VFR, scaled to 640x480 with 1:1 pixel aspect ratio`

When i interrupt the intro with a key press the colours are good and this is logged.
`DISPLAY: EGA 640x480 16-colour from 6-bit RGB palette (mode 13h) at 59.713 Hz VFR, scaled to 640x480 with 1:1 pixel aspect ratio`

The same is happening with `force_vga_single_scan` is `true` but the logs look like this.
`DISPLAY: VGA 640x480 16-colour from 6-bit RGB palette (mode 12h) at 59.713 Hz VFR, scaled to 640x478 with 1:0.996 pixel aspect ratio` (colours wrong)
`DISPLAY: EGA 640x240 16-colour from 6-bit RGB palette (mode 13h) at 59.713 Hz VFR, scaled to 640x478 with 1:1.99 pixel aspect ratio ` (colours good)

![intro_000](https://user-images.githubusercontent.com/62261787/236700253-8beccbd0-f342-424e-8737-93c2f699ecd0.png)

Pinball Illusions:
With `force_vga_single_scan` is `true` the pinball select screen is garbled and this is logged.
`DISPLAY: EGA 304x448 16-colour from 6-bit RGB palette (mode 0Dh) at 59.6 Hz VFR, scaled to 328x480 with 1:0.993 pixel aspect ratio`

![illusion_001](https://user-images.githubusercontent.com/62261787/236700258-f6c81c3f-ceb4-40ed-ab77-b8c3a52f04de.png)

kcgen commented:
Thanks for the help, @bitstrom. The latest push should have the palette correct in pinball fantasy. 

> With force_vga_single_scan is true the pinball select screen is garbled and this is logged.

Thanks. I need to rework this EGA setup portion to handle all these cases.. Back to the salt mines. ⛑️

johnnovak commented:
Very nice results and screenshots @kcgen, great stuff! It's not apparent on which screenshots you forced single scanlines, but I'm guessing on Amiga conversions such as the pinball games. Looks very cool!

bitstrom commented:
Thanks for the commitment @kcgen. Pinball Fantasies is working fine now.

I don't now if it is an error. I tested Duke Nukem 3D  with `VESA 320x200 8-bit (mode 150h)` and Stargunner with
`VESA 320x240 8-bit (mode 151h)` and they are not double-scanned with `force_vga_single_scan` is `false`.

johnnovak commented:
Great progress @kcgen. Almost good, but there are a few problems & regressions:

### `machine = svga_s3`

### Confusing RGB-palette bit-depth logging

These should be logged as "from 18-bit RGB palette" because a) technically that's the correct term (3 x 6-bit colour components), b) it's really easy to confuse this with the "6-bit RGB palette" used by EGA (3 x 2-bit colour components). In fact, when I first read "6-bit RGB" for VGA modes I thought the logging is wrong because of that...

Search for "6-bit RGB" and "18-bit RGB" here:
https://en.wikipedia.org/wiki/List_of_monochrome_and_RGB_color_formats

So these are wrong:
```
DISPLAY: EGA 640x400 16-colour from 6-bit RGB palette (mode 0Dh) at 70.086 Hz VFR, scaled to 2205x1654 with 1:1.2 pixel aspect ratio
```

```
DISPLAY: EGA 640x350 16-colour from 6-bit RGB palette (mode 10h) at 70.086 Hz VFR, scaled to 2205x1654 with 1:1.37 pixel aspect ratio
```

We could simply log "18-bit RGB" for all VGA modes... maybe. But then, things get murky as VBE 1.2 adds true 24-bit colour support, even to palettised modes:

> The new Set DAC Palette Control function enables the width of each primary (RGB) color element to be increased to 8 bits or more. The purpose is to enable 24-bit color definitions in the context of a palettized 256-color mode.
>
> This is the primary alternative to using paletteless Direct Color models. Applications may continue to address memory the same way they would in a 256-color mode, but the palette indices actually refer to 24-bit values rather than 18-bit values.

Source: http://ps-2.kev009.com/ohland/video/Volkman_VBE_12_Standard.html

So not sure if we want (or even can) be super precise about this to differentiate between 18-bit and 24-bit colour, but at least  using "18-bit colour" for all _palettised_ VGA modes would be an improvement as it would make things less ambiguous.

### Incorrect 24-bit VESA mode bit-depth logging

These should all say "24-bit", not "32-bit":

```
DISPLAY: VESA 640x480 32-bit (mode 112h) at 70.007 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
DISPLAY: VESA 800x600 32-bit (mode 115h) at 70.168 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
DISPLAY: VESA 1024x768 32-bit (mode 118h) at 70.052 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

### Incorrect 640x480 / 2-colour (mode 11h) logging

This is not monochrome; it's 2-colour. The two colours can be anything from the 18-bit RGB palette. Btw, I think only Hercules should be logged as "monochrome", and you're already logging "2-colour" correctly for all the other 2-colour modes on all adapters (VGA, CGA, EGA).

```
DISPLAY: MCGA 640x480 monochrome (mode 11h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

### Incorrect higher than 320x200 / 256-colour mode logging

Should say "256-colour from 18-bit RGB palette".

```
DISPLAY: VESA 640x480 8-bit (mode 101h) at 70.007 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

Same for 640x400:

```
DISPLAY: VESA 640x400 8-bit (mode 100h) at 69.755 Hz VFR, scaled to 2205x1654 with 1:1.2 pixel aspect ratio
```

Same for all the remaining 256-colour VESA modes; please test them in Deluxe Paint II or QPEG, SEA, etc.

### `machine = ega`

So yeah, the wording "6-bit RGA palette" here is correct—but only by a fluke 😄 (because you're currently logging "6-bit RGB palette" for VGA, and this mistakenly gets logged as the EGA modes on VGA...)

```
DISPLAY: EGA 320x200 16-colour from 6-bit RGB palette (mode 0Dh) at 59.923 Hz VFR, scaled to 2205x1654 with 1:1.2 pixel aspect ratio

```

So to be clear, we should log things differently on EGA and VGA for this mode (and all EGA modes):

- On EGA: `EGA 320x200 16-colour from 6-bit RGB palette (mode 0Dh)`
- On VGA: `EGA 320x200 16-colour from 18-bit RGB palette (mode 0Dh)`

---

Another problem: when you enter the 640x350 16-colour EGA _graphics_ mode (mode 10h) from the mode 03h text mode screen, nothing gets logged. I assume that's because the default 03h text mode is 640x350 on EGA. You can test this easily with Deluxe Paint II or early Legend adventures such as Spellcasting 101---nothing will get logged when the program switches from the default text mode to 640x350 (mode 10h).

---

To avoid playing a game of whack-a-mole, suggest testing _all_ modes quickly again in both Deluxe Paint II and QPEG to make sure you did not break something else. Deluxe Paint can handle Hercules, CGA, EGA, and (S)VGA adapters, and QPEG can only handle VGA, but it can display 15/16/24-bit VESA modes that Deluxe Paint can't (it can only do palettised modes).

kcgen commented:
@johnnovak , I made the colour and bit-depth logging fixes, but opted to drop the palette details; it was starting to feel overloaded with video-card internals - and even too much for me to look at on the log line.

Maybe there's a more elegant way to log it at the machine-level. For example, whenever we're in `machine = ega`, we could mention using a 6-bit / 64-colour palette, and 18-bit for VGA machines at one-time one startup, instead of repeating it all the graphical modes.

kcgen commented:
> Another problem: when you enter the 640x350 16-colour EGA graphics mode (mode 10h) from the mode 03h text mode screen, nothing gets logged.

Yes; unfortunately the logging happens downstream from the VGA side, and only on changes at the host level (if the SDL or OpenGL context / side needs refreshing).  I've added a small check for this, and it seems to be fixed:

```
2023-05-08 14:21:31.881 | DISPLAY: Text 640x350 16-colour (mode 03h) at 59.702 Hz CFR, scaled to 1440x1080 with 1:1.37 pixel aspect ratio

2023-05-08 14:21:42.362 | DISPLAY: EGA 640x350 16-colour (mode 10h) at 59.702 Hz CFR, scaled to 1440x1080 with 1:1.37 pixel aspect ratio

2023-05-08 14:21:46.307 | DISPLAY: Text 640x350 16-colour (mode 03h) at 59.702 Hz CFR, scaled to 1440x1080 with 1:1.37 pixel aspect ratio

2023-05-08 14:21:50.172 | DISPLAY: Text 640x350 16-colour (mode 02h) at 59.702 Hz CFR, scaled to 1440x1080 with 1:1.37 pixel aspect ratio

2023-05-08 14:22:07.866 | DISPLAY: EGA 640x350 16-colour (mode 10h) at 59.702 Hz CFR, scaled to 1440x1080 with 1:1.37 pixel aspect ratio
```

kcgen commented:
> Thanks for the commitment @kcgen. Pinball Fantasies is working fine now.
> 
> I don't now if it is an error. I tested Duke Nukem 3D with `VESA 320x200 8-bit (mode 150h)` and Stargunner with `VESA 320x240 8-bit (mode 151h)` and they are not double-scanned with `force_vga_single_scan` is `false`.

VESA modes are un-touched (except for logging now); I'd rather leave them be.

johnnovak commented:
> > Thanks for the commitment @kcgen. Pinball Fantasies is working fine now.
> > I don't now if it is an error. I tested Duke Nukem 3D with `VESA 320x200 8-bit (mode 150h)` and Stargunner with `VESA 320x240 8-bit (mode 151h)` and they are not double-scanned with `force_vga_single_scan` is `false`.
> 
> VESA modes are un-touched (except for logging now); I'd rather leave them be.

Yeah, maybe stop here. Let's leave something for the other guys to fix later 😎

johnnovak commented:
> @johnnovak , I made the colour and bit-depth logging fixes, but opted to drop the palette details; it was starting to feel overloaded with video-card internals - and even too much for me to look at on the log line.
> 
> Maybe there's a more elegant way to log it at the machine-level. For example, whenever we're in `machine = ega`, we could mention using a 6-bit / 64-colour palette, and 18-bit for VGA machines at one-time one startup, instead of repeating it all the graphical modes.

I agree with that; I almost suggested to drop the part about the palettes altogether, but then figured you must like it 😄 It was a getting a bit too much, agreed.

I'll re-test everything again later today with the recent changes.

johnnovak commented:
Almost there, only two more minor things 😅 

## `machine = svga_s3`

### 320x200 256-colour (mode 13h)

We log "VGA" for this one, but this mode was introduced on MCGA, actually. We
seem to follow a pattern to always use the video standard the mode was first
introduced on. So I think this should be "MCGA".

```
DISPLAY: VGA 640x400 256-colour (mode 13h) at 70.086 Hz VFR, scaled to 2205x1654 with 1:1.2 pixel aspect ratio
```

Btw, for 640x480 2-colour (mode 11h) we're correctly logging MCGA already:

```
DISPLAY: MCGA 640x480 2-colour (mode 11h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

### 640x480 16-colour (mode 12h)

That's not an EGA mode, should definitely log "VGA".

```
DISPLAY: EGA 640x480 16-colour (mode 12h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

johnnovak commented:
Almost there, only two more minor things 😅 

## `machine = svga_s3`

### 320x200 256-colour (mode 13h)

We log "VGA" for this one, but this mode was introduced on MCGA, actually. We
seem to follow a pattern to always use the video standard the mode was first
introduced on. So I think this should be "MCGA".

```
DISPLAY: VGA 640x400 256-colour (mode 13h) at 70.086 Hz VFR, scaled to 2205x1654 with 1:1.2 pixel aspect ratio
```

Btw, for 640x480 2-colour (mode 11h) we're correctly logging MCGA already:

```
DISPLAY: MCGA 640x480 2-colour (mode 11h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

### 640x480 16-colour (mode 12h)

That's not an EGA mode, should definitely log "VGA".

```
DISPLAY: EGA 640x480 16-colour (mode 12h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

johnnovak commented:
Almost there, only two more minor things 😅 

## `machine = svga_s3`

### 320x200 256-colour (mode 13h)

We log "VGA" for this one, but this mode was introduced on MCGA, actually. We
seem to follow a pattern to always use the video standard the mode was first
introduced on. So I think this should be "MCGA".

```
DISPLAY: VGA 640x400 256-colour (mode 13h) at 70.086 Hz VFR, scaled to 2205x1654 with 1:1.2 pixel aspect ratio
```

Btw, for 640x480 2-colour (mode 11h) we're correctly logging MCGA already:

```
DISPLAY: MCGA 640x480 2-colour (mode 11h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

### 640x480 16-colour (mode 12h)

That's not an EGA mode, should definitely log "VGA".

```
DISPLAY: EGA 640x480 16-colour (mode 12h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

johnnovak commented:
Almost there, only two more minor things 😅 

## `machine = svga_s3`

### 320x200 256-colour (mode 13h)

We log "VGA" for this one, but this mode was introduced on MCGA, actually. We
seem to follow a pattern to always use the video standard the mode was first
introduced on. So I think this should be "MCGA".

```
DISPLAY: VGA 640x400 256-colour (mode 13h) at 70.086 Hz VFR, scaled to 2205x1654 with 1:1.2 pixel aspect ratio
```

Btw, for 640x480 2-colour (mode 11h) we're correctly logging MCGA already:

```
DISPLAY: MCGA 640x480 2-colour (mode 11h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

### 640x480 16-colour (mode 12h)

That's not an EGA mode, should definitely log "VGA".

```
DISPLAY: EGA 640x480 16-colour (mode 12h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

johnnovak commented:
Almost there, only two more minor things 😅 

## `machine = svga_s3`

### 320x200 256-colour (mode 13h)

We log "VGA" for this one, but this mode was introduced on MCGA, actually. We
seem to follow a pattern to always use the video standard the mode was first
introduced on. So I think this should be "MCGA".

```
DISPLAY: VGA 640x400 256-colour (mode 13h) at 70.086 Hz VFR, scaled to 2205x1654 with 1:1.2 pixel aspect ratio
```

Btw, for 640x480 2-colour (mode 11h) we're correctly logging MCGA already:

```
DISPLAY: MCGA 640x480 2-colour (mode 11h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

### 640x480 16-colour (mode 12h)

That's not an EGA mode, should definitely log "VGA".

```
DISPLAY: EGA 640x480 16-colour (mode 12h) at 59.94 Hz VFR, scaled to 2205x1654 with 1:1 pixel aspect ratio
```

kcgen commented:
> ### 320x200 256-colour (mode 13h)
> 
> We log "VGA" for this one, but this mode was introduced on MCGA, actually. We seem to follow a pattern to always use the video standard the mode was first introduced on. So I think this should be "MCGA".

Got it, and specific the rigid 320x200 (or double-scanned) resolution.

For example, here's Jurassic Park, which uses all kinds of mode 13 resolutions:

```
DISPLAY: MCGA 640x400 256-colour (mode 13h)
...
DISPLAY: VGA 640x408 256-colour (mode 13h)
...
DISPLAY: VGA 640x370 256-colour (mode 13h)
```


> ### 640x480 16-colour (mode 12h)
> 
> That's not an EGA mode, should definitely log "VGA".

Fixed - thanks!

```
DISPLAY: VGA 640x480 16-colour (mode 12h) at 59.94 Hz
```

kcgen commented:
Took a final pass through; all commits build, the modes are check-out and log OK.  Will merge - thank for all the testing and feedback @johnnovak , @GranMinigun , and @bitstrom.
